packages/react-tabs/src/components/Tab/useTabAnimatedIndicator.ts

```diff
@@ -0,0 +1,180 @@
+      bottom: `calc(var(${tabIndicatorCssVars_unstable.endOffsetVar}) + ${tabIndicatorPadding.mediumVertical})`,
+    },
+  },
+  smallHorizontal: {
+    ':after': {
+      left: `calc(var(${tabIndicatorCssVars_unstable.startOffsetVar}) + ${tabIndicatorPadding.smallHorizontal})`,
+      right: `calc(var(${tabIndicatorCssVars_unstable.endOffsetVar}) + ${tabIndicatorPadding.smallHorizontal})`,
+    },
+  },
+  smallVertical: {
+    ':after': {
+      top: `calc(var(${tabIndicatorCssVars_unstable.startOffsetVar}) + ${tabIndicatorPadding.smallVertical})`,
+      bottom: `calc(var(${tabIndicatorCssVars_unstable.endOffsetVar}) + ${tabIndicatorPadding.smallVertical})`,
+    },
+  },
+});
+
+const calculateTabRect = (element: HTMLElement) => {
+  if (element) {
+    const parentRect = element.parentElement?.getBoundingClientRect() || { x: 0, y: 0, width: 0, height: 0 };
+    const tabRect = element.getBoundingClientRect();
+
+    return {
+      x: tabRect.x - parentRect.x,
+      y: tabRect.y - parentRect.y,
+      width: tabRect.width,
+      height: tabRect.height,
+    };
+  }
+  return undefined;
+};
+
+const getRegisteredTabRect = (registeredTabs: Record<string, TabRegisterData>, value?: TabValue) => {
+  const element = value ? registeredTabs[JSON.stringify(value)]?.ref.current : undefined;
+  return element ? calculateTabRect(element) : undefined;
+};
+
+/**
+ * Adds additional styling to the active tab selection indicator to create a sliding animation.
+ */
+export const useTabAnimatedIndicatorStyles_unstable = (state: TabState): TabState => {
+  const { disabled, selected, vertical } = state;
+
+  const activeIndicatorStyles = useActiveIndicatorStyles();
+  const [lastAnimatedFrom, setLastAnimatedFrom] = React.useState<TabValue>();
+  const [setAnimateTimeout] = useTimeout();
+  const getRegisteredTabs = useContextSelector(TabListContext, ctx => ctx.getRegisteredTabs);
+
+  let indicatorStartOffset = 0;
+  let indicatorEndOffset = 0;
+  if (selected) {
+    const { previousSelectedValue, selectedValue, registeredTabs } = getRegisteredTabs();
+    const previousSelectedTabRect = getRegisteredTabRect(registeredTabs, previousSelectedValue);
+    const selectedTabRect = getRegisteredTabRect(registeredTabs, selectedValue);
+
+    if (
+      selectedTabRect &&
+      previousSelectedTabRect &&
+      previousSelectedValue &&
+      lastAnimatedFrom !== previousSelectedValue
+    ) {
+      indicatorStartOffset = vertical
+        ? previousSelectedTabRect.y - selectedTabRect.y
+        : previousSelectedTabRect.x - selectedTabRect.x;
+
+      indicatorEndOffset = vertical
+        ? selectedTabRect.y + selectedTabRect.height - (previousSelectedTabRect.y + previousSelectedTabRect.height)
+        : selectedTabRect.x + selectedTabRect.width - (previousSelectedTabRect.x + previousSelectedTabRect.width);
+
+      setAnimateTimeout(() => {
+        setLastAnimatedFrom(previousSelectedValue);
+      }, 1);
```

Comment:
- `setAnimateTimeout()` looks like `requestAnimationFrame` actually, is there need to use `setTimeout`?
- Did you consider to run this in effect somehow? React 17/18 in [Strict Mode](https://reactjs.org/blog/2022/03/29/react-v18.html#new-strict-mode-behaviors) runs rendering twice (https://codesandbox.io/s/new-fast-891888?file=/src/index.js). So it might be safer to move some work in effects.

Reply:
I updated to useEffect.  In the FLIP pattern it is common to use a timeout to set state and create a tick() like function between progressive animation stages, but useEffect can do the same thing if the offsets are state.